libraries/AP_RangeFinder/AP_RangeFinder_JRE_Serial.cpp

```diff
@@ -0,0 +1,110 @@
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_RangeFinder_JRE_Serial.h"
+#include <AP_Math/AP_Math.h>
+
+#if AP_RANGEFINDER_JRE_SERIAL_ENABLED
+
+#define FRAME_HEADER_1   'R'    // 0x52
+#define FRAME_HEADER_2   'A'    // 0x41
+
+#define DIST_MAX_CM 50000
+#define OUT_OF_RANGE_ADD_CM   100
+
+bool AP_RangeFinder_JRE_Serial::get_reading(float &reading_m)
+{
+    // uart instance check
+    if (uart == nullptr) {
+        return false;  // not update
+    }
+    uint32_t n = uart->available();
+    if (n == 0) {
+        return false;
+    }
+
+    uint16_t valid_count = 0;   // number of valid readings
+    uint16_t invalid_count = 0; // number of invalid readings
+    float sum = 0;
+    float reading_cm;
+    // max distance the sensor can reliably measure - read from parameters
+    const int16_t distance_cm_max = max_distance_cm();
+
+    // buffer read
+    const ssize_t num_read = uart->read(read_buff, ARRAY_SIZE(read_buff));
+    read_buff_idx = 0; // read_buff start data index
+    while (read_buff_idx < num_read) {
+
+        if (data_buff_idx == 0) { // header first byte check
+            // header data search
+            for (; read_buff_idx < num_read; read_buff_idx++) {
+                if (read_buff[read_buff_idx] == FRAME_HEADER_1) {
+                    data_buff[0] = FRAME_HEADER_1;
+                    data_buff_idx = 1; // next data_buff
+                    read_buff_idx++;   // next read_buff
+                    break;
+                }
+            }
+
+        } else if (data_buff_idx == 1) { // header second byte check
+            if (read_buff[read_buff_idx] == FRAME_HEADER_2) {
+                data_buff[1] = FRAME_HEADER_2;
+                data_buff_idx = 2;    // next data_buff
+            } else {
+                data_buff_idx = 0;    // data index clear
+            }
+            read_buff_idx++;          // next read_buff
+
+        } else { // data set
+            if (data_buff_idx >= ARRAY_SIZE(data_buff)) {  // 1 data set complete
+                // crc check
+                uint16_t crc = crc16_ccitt_r(data_buff, ARRAY_SIZE(data_buff) - 2, 0xffff, 0xffff);
+                if ((((crc>>8) & 0xff) == data_buff[15]) && ((crc & 0xff) == data_buff[14])) {
+                    // status check
+                    if (data_buff[13] & 0x02) { // NTRK
+                        invalid_count++;
+                    } else { // TRK
+                        reading_cm = data_buff[4] * 256 + data_buff[5];
+                        if (reading_cm < distance_cm_max) {
```

Comment:
I think we should remove this "reading_cm < distance_cm_max" check.  Imagine this scenarios:

- max distance is set to 50m
- two distances arrive, 75m and 45m
- 75m will be shortened to 50m, the average will become 47.5m which will be considered valid and will be used by the autopilot (for surface tracking for example)

Instead what we should do is average 75 and 45 to arrive at 60m which will then be determined (by higher level code) to be out-of-range-high and the value will not be used.

Reply:
I corrected it.